### PR TITLE
Should not retry CQL ServerErrors for large key values

### PR DESCRIPTION
**Pull request by geekmug**
For Cassandra, the QueryValidationException was wrapped in an ExecutionException causing the EXCEPTION_MAPPER to not match, but for SycllaDB, the failure is reported as a general ServerError, so we can only match on this condition by inspectng the detail message.

Signed-off-by: Scott Dial <geekmug@google.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

File: janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java [hunk 1]
```diff
@@ -117,9 +121,49 @@ public class CQLKeyColumnValueStore implements KeyColumnValueStore {
...
+                    ServerError error = (ServerError) throwable;
+                    if (error.getMessage().startsWith("Key size too large:")) {
+                        return true;
+                    }
+                    return false;
+                }
+                return false;
+            });
+        });
+    }
```
Comment (porunov) [inline janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:150]: As Codacy commented in this PR.
1) it's better to declare all methods after all fields are declared. I.e. it's better to move this method down after all fields are declared.
2) It's possible to make boolean statements easier without if / else statements. I.e. something like below:
```
private static boolean isKeySizeTooLarge(Throwable cause) {
    AllNodesFailedException failed = ExceptionUtils.throwableOfType(cause, AllNodesFailedException.class);
    if (failed == null) {
        return false;
    }
    return failed.getAllErrors().values().stream().anyMatch(errors ->
        errors.stream().anyMatch(error -> (error instanceof ServerError)
            && error.getMessage().startsWith("Key size too large:")));
}
```

**Review by porunov — APPROVED** [review]
I have only a single nitpick about `isKeySizeTooLarge` as Codacy commented. Otherwise the fix looks good to me. Thank you @geekmug !

File: janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java [hunk 2]
```diff
@@ -117,9 +121,37 @@ public class CQLKeyColumnValueStore implements KeyColumnValueStore {
...
+     * This type of ServerError is a permanent failure, since the same query with
+     * the same value will always produce this error. Properly treating this as a
+     * permanent failure avoids excessive retrying of bad CQL queries and allows
+     * upstream code to properly report the exception.
+     */
+    private static boolean isKeySizeTooLarge(Throwable cause) {
+        AllNodesFailedException failed = ExceptionUtils.throwableOfType(cause, AllNodesFailedException.class);
+        return failed != null && failed.getAllErrors().values().stream().anyMatch(errors ->
+            errors.stream().anyMatch(error -> (error instanceof ServerError)
+                && error.getMessage().startsWith("Key size too large:")));
```
Comment (geekmug) [inline janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:153]: For reference, this originates from this bit of code in ScyllaDB: https://github.com/scylladb/scylladb/blob/8d7e35caefecc053ba5495c519cc270632c6db1a/compound.hh#L106

**Comment by geekmug** [comment]
@porunov Thanks for the feedback and code review! Your changes looked good to me and I re-tested everything locally. Thanks!

**Review by porunov — APPROVED** [review]

**Comment by porunov** [comment]
Merging by following lazy consensus

**Pull request merged**